src/sources/timer/mod.rs

```diff
@@ -9,15 +9,39 @@
 
 use std::cell::RefCell;
 use std::collections::BinaryHeap;
-use std::sync::{
-    atomic::{AtomicBool, Ordering},
-    Arc, Mutex,
-};
+use std::sync::{Arc, Mutex};
 use std::time::{Duration, Instant};
 
-use super::ping::{make_ping, PingError, PingSource};
 use crate::{EventSource, Poll, PostAction, Readiness, Token, TokenFactory};
 
+#[cfg(target_os = "linux")]
+mod timerfd;
+#[cfg(target_os = "linux")]
+use timerfd::{TimerScheduler, TimerSource};
+
+#[cfg(any(
+    target_os = "dragonfly",
+    target_os = "freebsd",
+    target_os = "netbsd",
+    target_os = "openbsd",
+    target_os = "macos"
+))]
+mod threaded;
+
+#[cfg(any(
+    target_os = "dragonfly",
+    target_os = "freebsd",
+    target_os = "netbsd",
+    target_os = "openbsd",
+    target_os = "macos"
+))]
+use threaded::{TimerScheduler, TimerSource};
+
+/// An error arising from processing events for a timer.
+#[derive(thiserror::Error, Debug)]
+#[error(transparent)]
+pub struct TimerError(Box<dyn std::error::Error + Sync + Send>);
```

Comment:
Why not make `TimerError` a platform-dependent type like `TimerScheduler` and `TimerSource`, rather than using boxing?

Reply:
So you can write cross platform timer handling without even knowing about it? If you have different types in public API the code will change downstream for bsd/linux however if you use common error type they won't even know that something is different? If it was a calloop internal type I'd go with it.